ecs-agent/credentials/providers/rotating_shared_credentials_provider_v2.go

```diff
@@ -0,0 +1,80 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package providers
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+)
+
+// RotatingSharedCredentialsProviderV2 is a provider that retrieves credentials via the
+// shared credentials provider, and adds the functionality of expiring and re-retrieving
+// those credentials from the file.
+// TODO (@tiffwang): Remove V2 suffix after the credentials package is
+// fully migrated to aws-sdk-go-v2.
+type RotatingSharedCredentialsProviderV2 struct {
+	RotationInterval time.Duration
+	profile          string
+	file             string
+}
+
+// NewRotatingSharedCredentials returns a rotating shared credentials provider
+// with default values set.
+func NewRotatingSharedCredentialsProviderV2() *RotatingSharedCredentialsProviderV2 {
+	var credentialProfile = DEFAULT_CREDENTIAL_PROFILE
+	if alternateCredentialProfile := os.Getenv(ALTERNATE_CREDENTIAL_PROFILE_ENV_VAR); alternateCredentialProfile != "" {
+		logger.Info(fmt.Sprintf("Overriding %s credential profile; using: %s.", DEFAULT_CREDENTIAL_PROFILE, alternateCredentialProfile))
+		credentialProfile = alternateCredentialProfile
+	}
+
+	return &RotatingSharedCredentialsProviderV2{
+		RotationInterval: defaultRotationInterval,
+		profile:          credentialProfile,
+		file:             defaultRotatingCredentialsFilename,
+	}
+}
+
+// Retrieve will use the given filename and profile and retrieve AWS credentials.
+func (p *RotatingSharedCredentialsProviderV2) Retrieve(ctx context.Context) (aws.Credentials, error) {
+	sharedConfig, err := config.LoadSharedConfigProfile(ctx, p.profile, func(option *config.LoadSharedConfigOptions) {
+		option.CredentialsFiles = []string{p.file}
+	})
+	credentials := sharedConfig.Credentials
+	credentials.Source = RotatingSharedCredentialsProviderName
+	if err != nil {
+		return credentials, err
+	}
+
+	credentials.CanExpire = true
+	credentials.Expires = time.Now().Add(p.RotationInterval)
+	logger.Info(fmt.Sprintf("Successfully got instance credentials from file %s. %s",
+		p.file, credentialsToString(credentials)))
```

Comment:
If I'm understanding correctly, in V2 `RotatingSharedCredentialsProviderV2.Retrieve` function calls `config.LoadSharedConfigProfile` function which, if credential file(s) are not set, defaults to using file at location `$HOME/.aws/credentials` (if it is present).

Meanwhile in V1, `RotatingSharedCredentialsProvider.Retrieve` function calls `SharedCredentialsProvider.Retrieve` function which, if credential file is not set, additionally checks if `AWS_SHARED_CREDENTIALS_FILE` environment variable is set (and uses its value as the file location if so) before defaulting to using file at location `$HOME/.aws/credentials` (if it is present).

This change in behavior is probably not a concern right now since it seems the value stored in `defaultRotatingCredentialsFilename` is always set as credential file (see L51) both before (as of commit SHA 9494fb353ff114a68f7406b454455cc5bf4768af) and after this change. 

However:
 1. Could `RotatingSharedCredentialsProviderV2.file` and/or `defaultRotatingCredentialsFilename` possibly be an empty string in the future? 
 1. Have `RotatingSharedCredentialsProvider.sharedCredentialsProvider.Filename` and/or `defaultRotatingCredentialsFilename` ever been an empty string in the past?  

If the answer to both of these is yes, then would we want to address this change in behavior (i.e., in the event credential file(s) are not set, would we want to ensure `AWS_SHARED_CREDENTIALS_FILE` environment variable is checked before defaulting to use file at location `$HOME/.aws/credentials`)?

Reply:
I think it's extremely unlikely for the filename to be empty. https://github.com/aws/amazon-ecs-agent/pull/2849 and https://github.com/aws/amazon-ecs-agent/pull/3184 are the only 2 PRs that touch the filename, but the filename is still `/rotatingcreds/credentials` (for Linux, at least) in both PRs.

Additionally, changing the filename or letting users override it with `AWS_SHARED_CREDENTIALS_FILE` would potentially break rotating creds, even without the SDK migration. Rotating creds are used with ECS-A, which is the only time we mount `/root/.aws` on the host to `/rotatingcreds` on the container. So if the rotating creds provider were configured to read creds outside of `/rotatingcreds`, it wouldn't find any. Supporting this would require additional changes to ecs-init and agent.
https://github.com/aws/amazon-ecs-agent/blob/ea4ffcdd5a69b36c5a7216938949fae0b26d2b0f/ecs-init/docker/docker.go#L463-L466